### PR TITLE
Tremblp/maya 106088/arbitrary stack 00

### PR DESCRIPTION
**Pull request by ppt-adsk**
Added Maya fallback transform op stack, and added common API transform op stack handling.

### CODE REVIEW

File: lib/mayaUsd/ufe/RotationUtils.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+//------------------------------------------------------------------------------
+// Helper functions for dealing with rotations.
+//------------------------------------------------------------------------------
+
+inline double TO_DEG(double a) { return a * 180.0 / 3.141592654; }
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/RotationUtils.h:35]: Is it used anywhere else than RotationUtils.cpp? If not, consider moving to an anonymous namespace.

File: lib/mayaUsd/ufe/UsdTransform3dFallbackMayaXformStack.cpp [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    }
+
+    xformable.SetXformOpOrder(newOrder, resetsXformStack);
+}
+
+Ufe::Transform3d::Ptr createTransform3d(const Ufe::SceneItem::Ptr& item)
+{
+    UsdSceneItem::Ptr usdItem = std::dynamic_pointer_cast<UsdSceneItem>(item);
+#if !defined(NDEBUG)
+    assert(usdItem);
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/UsdTransform3dFallbackMayaXformStack.cpp:173]: Can we use TF_CODING_ERROR/TF_FATAL_ERROR instead?

File: lib/mayaUsd/ufe/UsdTransform3dFallbackMayaXformStack.cpp [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    // is well, from the first fallback op onwards, we have a sub-stack that
+    // matches the fallback Maya transform stack.
+    std::vector<UsdGeomXformOp> candidateOps;
+    candidateOps.reserve(std::distance(i, xformOps.cend()));
+    std::copy(i, xformOps.cend(), std::back_inserter(candidateOps));
+
+    // We're the last handler in the chain of responsibility: if the candidate
+    // ops support the Maya transform stack, create a Maya transform stack
+    // interface for it, otherwise no further handlers to delegate to, so fail.
+    return MatchingSubstack(candidateOps) ? UsdTransform3dFallbackMayaXformStack::create(usdItem)
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/UsdTransform3dFallbackMayaXformStack.cpp:212]: Do we have examples of stacks that will fail?

File: lib/mayaUsd/ufe/UsdTransform3dFallbackMayaXformStack.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+//! \brief Factory to create the fallback Transform3d interface object.
+//
+// Since this is the fallback Transform3d handler, it is the final handler in
+// the chain of responsibility.
+
+class MAYAUSD_CORE_PUBLIC UsdTransform3dFallbackMayaXformStackHandler
+    : public Ufe::Transform3dHandler
+{
+public:
+    typedef std::shared_ptr<UsdTransform3dFallbackMayaXformStackHandler> Ptr;
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/UsdTransform3dFallbackMayaXformStack.h:100]: At one point we will get clang-tidy to modernize usage of typedef - https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-using.html

File: lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp [hunk 5]
```diff
@@ -255,34 +170,122 @@ Ufe::Transform3d::Ptr createTransform3d(
...
+    const UsdTimeCode _readTime;
+    const UsdTimeCode _writeTime;
+    VtValue           _prevOpValue;
+    VtValue           _newOpValue;
+    TfToken           _attrName;
+    UsdGeomXformOp    _op;
+    OpFunc            _opFunc;
 
 public:
+    struct State
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp:182]: This needs much more documentation and preferably be taken out into its own class.

File: lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp [hunk 6]
```diff
@@ -255,34 +170,122 @@ Ufe::Transform3d::Ptr createTransform3d(
...
+        , _opFunc(std::move(opFunc))
     {
     }
 
-    // Have separate execute override which is value-only, as default execute()
-    // calls redo, and undo / redo will eventually not be value only and will
-    // support propertySpec / primSpec management.
-    void execute() override { setValue(fNewOpValue); }
+    // Ufe::UndoableCommand overrides.
+    void execute() override { handleSet(_newOpValue); }
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp:286]: why do we need this handleSet method? for undo you call the state directly.

File: lib/mayaUsd/ufe/wrapUtils.cpp [hunk 7]
```diff
@@ -76,28 +77,7 @@ std::string stagePath(UsdStageWeakPtr stage)
...
-    // separator is '|', so use a map to get the corresponding UFE run-time ID.
-    Ufe::Path                        path;
-    static std::map<char, Ufe::Rtid> sepToRtid = { { '/', ufe::getUsdRunTimeId() }, { '|', 1 } };
-    for (std::size_t i = 0; i < segmentStrings.size(); ++i) {
-        const auto& segmentString = segmentStrings[i];
-        char        sep = segmentString[0];
-        path = path + Ufe::PathSegment(segmentString, sepToRtid.at(sep), sep);
-    }
-    return ufe::ufePathToPrim(path);
+    return ufe::ufePathToPrim(Ufe::PathString::path(ufePathString));
```
Comment (kxl-adsk) [inline lib/mayaUsd/ufe/wrapUtils.cpp:80]: Curious - is this related to the change or simply cleanup?

File: test/lib/ufe/testTransform3dChainOfResponsibility.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+import mayaUsd.ufe
+
+import mayaUtils
+import ufe
+
+from pxr import UsdGeom, Vt
+
+import unittest
+
+class Transform3dChainOfResponsibilityTestCase(unittest.TestCase):
```
Comment (kxl-adsk) [inline test/lib/ufe/testTransform3dChainOfResponsibility.py:30]: This is great. Do we have a way to change the chain of responsibility from python at runtime?

**Review by kxl-adsk — APPROVED** [review]
Lots of good work, @ppt-adsk !

File: lib/mayaUsd/ufe/wrapUtils.cpp (hunk 7 above)
Comment (ppt-adsk) [reply lib/mayaUsd/ufe/wrapUtils.cpp:80]: Cleanup.  Saw this, couldn't resist.

File: test/lib/ufe/testTransform3dChainOfResponsibility.py (hunk 8 above)
Comment (ppt-adsk) [reply test/lib/ufe/testTransform3dChainOfResponsibility.py:30]: No, no bindings.

File: lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp (hunk 6 above)
Comment (ppt-adsk) [reply lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp:286]: Turns out command derived classes call this, and I don't want to expose them to the base class state, which is an implementation detail.
